homeassistant/components/combined_energy/sensor.py

```diff
@@ -0,0 +1,270 @@
+"""Support for getting collected information from Combined Energy."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import Generator, Sequence
+from datetime import datetime
+
+from combined_energy import CombinedEnergy
+from combined_energy.models import Device, DeviceReadings, Installation
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import (
+    DATA_API_CLIENT,
+    DATA_INSTALLATION,
+    DATA_LOG_SESSION,
+    DOMAIN,
+    SENSOR_DESCRIPTION_CONNECTED,
+    SENSOR_DESCRIPTIONS,
+)
+from .coordinator import (
+    CombinedEnergyConnectivityDataService,
+    CombinedEnergyLogSessionService,
+    CombinedEnergyReadingsDataService,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors."""
+
+    api: CombinedEnergy = hass.data[DOMAIN][entry.entry_id][DATA_API_CLIENT]
+    installation: Installation = hass.data[DOMAIN][entry.entry_id][DATA_INSTALLATION]
+
+    # Initialise services
+    connection = CombinedEnergyConnectivityDataService(hass, api)
+    log_session = CombinedEnergyLogSessionService(hass, api)
+    readings = CombinedEnergyReadingsDataService(hass, api)
+    for service in (connection, log_session, readings):
+        service.async_setup()
+        await service.coordinator.async_refresh()
+
+    # Store log session into Data
+    hass.data[DOMAIN][entry.entry_id][DATA_LOG_SESSION] = log_session
+
+    # Build entity list
+    sensor_factory = CombinedEnergyReadingsSensorFactory(hass, installation, readings)
+    entities: list[CombinedEnergyReadingsSensor | CombinedEnergyConnectedSensor] = list(
+        sensor_factory.entities()
+    )
+    # Insert connected as the first sensor
+    entities.insert(0, CombinedEnergyConnectedSensor(entry.title, connection))
+
+    async_add_entities(entities)
+
+
+class CombinedEnergyConnectedSensor(CoordinatorEntity, BinarySensorEntity):
```

Comment:
CoordinatorEntity[YourCoordinatorName] is ideal, but I'm not sure how that works with your services system you built

Reply:
Also this is a binary sensor, you can't have this in here/ your initial PR. 

PRs are limited to one platform, aka only Sensors or only Binary Sensors

Reply:
This is an important piece of state information that indicates if the monitoring hardware is connected to the Combined Energy monitoring platform. I don't have a view on how to exclude this entity, would prefer to defer to a maintainer's opinion.

Reply:
Absolutely feel free to defer for their opinion, but from what I've seen it is a pretty strong rule they hold to. Typically it is also in a seperate file binary_sensor.py where you can set up the BinarySensor platform separately from the non-binary sensors.

But absolutely, feel free to defer to them